File: Java/src/_0538_Convert_BST_to_Greater_Tree/Solution.java
package _0538_Convert_BST_to_Greater_Tree;

/**
 * Date: 2020/9/21 8:14
 * Content:
 * BST 中序遍历 是 有序的
 * 反向中序遍历：左根右 --> 右根左
 */
public class Solution {
    int sum = 0;

    // 将以 root 为根的 BST 转为累加树
    public TreeNode convertBST(TreeNode root) {
        if (root != null) {
            convertBST(root.right);
            sum += root.val;
            root.val = sum;
            convertBST(root.left);
        }
        return root;
    }

    public static void main(String[] args) {
        int nu = Integer.MIN_VALUE;  // 用Integer.MIN_VALUE表示二叉树 null
        int[] nums = new int[]{
                5,
                3, 13,
                nu, 4, 7, 16
        };

        MyTree myTree = new MyTree(nums);
        TreeNode root = myTree.getRoot();
        (new Solution()).convertBST(root);
        System.out.println(myTree.levelOrder());
    }
}

File: Python/_0538_Convert_BST_to_Greater_Tree/Solution.py
# coding=utf-8
# Date: 2021/7/16 16:57
from MyTree import *


# 执行用时：84 ms, 在所有 Python3 提交中击败了24.67%的用户
# 内存消耗：17.1 MB, 在所有 Python3 提交中击败了51.92%的用户
class Solution:
    def __init__(self):
        self.sum = 0

    def convertBST(self, root: TreeNode) -> TreeNode:
        self.dfs(root)
        return root

    def dfs(self, root):
        if root is not None:
            self.dfs(root.right)

            self.sum += root.val
            root.val = self.sum

            self.dfs(root.left)


if __name__ == "__main__":
    nu = None
    nums = [
        5,
        3, 13,
        nu, 4, 7, 16
    ]
    root = MyTree(nums).getRoot()
    Solution().convertBST(root)
    print(MyTree().levelOrder(root))  # [[41], [48, 29], [45, 36, 16]]
